### PR TITLE
Test of compact points node

### PR DESCRIPTION
**Pull request by nebraszka**

### CODE REVIEW

File: test/src/graph/nodes/CompactPointsNodeTest.cpp [hunk 1]
```diff
@@ -1,24 +1,135 @@
...
+
+        switch(field) {
+            case XYZ_F32: {
+                std::vector<::Field<XYZ_F32>::type> outData;
+                outData.resize(outCount);
+                EXPECT_RGL_SUCCESS(rgl_graph_get_result_data(compactNode, field, outData.data()));
+                for (int i = 0; i < outCount; ++i) {
+                    EXPECT_NEAR(expectedHitPoints.at(i).xyz[0], outData.at(i)[0], EPSILON_F);
+                    EXPECT_NEAR(expectedHitPoints.at(i).xyz[1], outData.at(i)[1], EPSILON_F);
+                    EXPECT_NEAR(expectedHitPoints.at(i).xyz[2], outData.at(i)[2], EPSILON_F);
```
Comment (prybicki) [inline test/src/graph/nodes/CompactPointsNodeTest.cpp:111]: Consider using `EXPECT_EQ`. I'd expect the data to remain unchanged bit-wise.

Comment (prybicki) [reply test/src/graph/nodes/CompactPointsNodeTest.cpp:111]: If it does not pass, consider comparing the data from the parent node. 
You can add a field filled with std::iota to help find indices to parent data after compaction.

File: test/src/graph/nodes/CompactPointsNodeTest.cpp [hunk 2]
```diff
@@ -1,24 +1,135 @@
...
+                    EXPECT_EQ(1, outData.at(i));
+                }
+                break;
+            }
+            case INTENSITY_F32: {
+                std::vector<::Field<INTENSITY_F32>::type> outData;
+                outData.resize(outCount);
+                EXPECT_RGL_SUCCESS(rgl_graph_get_result_data(compactNode, field, outData.data()));
+                for (int i = 0; i < outCount; ++i) {
+                    EXPECT_NEAR(expectedHitPoints.at(i).intensity, outData.at(i), EPSILON_F);
```
Comment (prybicki) [inline test/src/graph/nodes/CompactPointsNodeTest.cpp:129]: ditto

**Review by prybicki — CHANGES_REQUESTED** [review]
Good code and appropriate tests. A few things to be addressed before merging. Most importantly - repeatability (add printing seed). 

Here are a few more ideas to test:
- Pass zero numbers to compact
- Try running compact without any input node
- Run node multiple times (both on same and changing data)
- Try running compact on input without IS_HIT field

File: test/src/graph/nodes/CompactPointsNodeTest.cpp [hunk 3]
```diff
@@ -1,24 +1,216 @@
...
+{
+    int pointsCount = GetParam();
+    prepareAndRunGraph(pointsCount, HitPointDensity::ALL_NON_HIT);
+
+    int32_t hitpointCount, pointSize;
+    ASSERT_RGL_SUCCESS(rgl_graph_get_result_size(compactNodes.at(0), RGL_FIELD_XYZ_F32, &hitpointCount, &pointSize));
+    EXPECT_EQ(hitpointCount, 0);
+
+    // Check if the contents of outData have changed (they should not have)
+    std::vector<::Field<XYZ_F32>::type> outData { { 2.0f, 3.0f, 5.0f }, { 7.0f, 11.0f, 13.0f }, { 17.0f, 19.0f, 23.0f } };
```
Comment (PiotrMrozik) [inline test/src/graph/nodes/CompactPointsNodeTest.cpp:76]: I was looking for 420 or 2137.

File: test/src/graph/nodes/CompactPointsNodeTest.cpp [hunk 4]
```diff
@@ -1,24 +1,216 @@
...
-    EXPECT_RGL_SUCCESS(rgl_node_points_compact(&compactNode));
+    RGLTestPointsIsHitGenerator* pointsGenerator = dynamic_cast<RGLTestPointsIsHitGenerator*>(generator.get());
+    std::vector<TestPointIsHit> expectedHitPoints = pointsGenerator->separateHitPoints();
+    std::vector<TestPointIsHit> returnedPoints = getResults(compactNodes.at(compactNodes.size() - 1));
+    verifyResults(returnedPoints, expectedHitPoints);
 }
+
+TEST_F(CompactPointsNodeTest, should_not_change_result_when_multiple_compactions_applied_to_changing_data)
+{
+    int pointsCount = 1000;
```
Comment (PiotrMrozik) [inline test/src/graph/nodes/CompactPointsNodeTest.cpp:157]: ditto

File: test/src/graph/nodes/CompactPointsNodeTest.cpp [hunk 5]
```diff
@@ -1,24 +1,216 @@
...
+}
+
+TEST_F(CompactPointsNodeTest, should_warn_when_empty_point_cloud)
+{
+    rgl_node_t emptyPointCloudOutputNode = simulateEmptyPointCloudOutputNode();
+
+    ASSERT_RGL_SUCCESS(rgl_node_points_compact(&compactNodes.at(0)));
+    ASSERT_RGL_SUCCESS(rgl_graph_node_add_child(emptyPointCloudOutputNode, compactNodes.at(0)));
+
+    //  TODO: Causes Segmentation fault
```
Comment (PiotrMrozik) [inline test/src/graph/nodes/CompactPointsNodeTest.cpp:199]: Create an issue on out board pls.

**Review by PiotrMrozik — CHANGES_REQUESTED** [review]

File: test/src/graph/nodes/CompactPointsNodeTest.cpp [hunk 6]
```diff
@@ -1,23 +1,229 @@
...
+    };
+	auto outDataCopy = outData;
+
+	ASSERT_RGL_SUCCESS(rgl_graph_get_result_data(compactNodes.at(0), RGL_FIELD_XYZ_F32, outData.data()));
+	ASSERT_EQ(outData.size(), outDataCopy.size());
+	for (int i = 0; i < outData.size(); ++i) {
+		EXPECT_EQ(outData.at(i)[0], outDataCopy.at(i)[0]);
+		EXPECT_EQ(outData.at(i)[1], outDataCopy.at(i)[1]);
+		EXPECT_EQ(outData.at(i)[2], outDataCopy.at(i)[2]);
+	}
```
Comment (prybicki) [inline test/src/graph/nodes/CompactPointsNodeTest.cpp:133]: 1. Add a check that `rgl_graph_get_result_size` is 0. 
2. Add a comment that you are verifying that no overwrite is happening, it's not obvious at first sight.

**Review by prybicki — APPROVED** [review]
Non-critical comments that may be addressed in future PRs. Tests are good enough to merge as of now.

General hint - I think that `CompactPointsNodeTest` is being overused. It is meant to be a fixture, i.e. set up some preconditions that are required in all tests. This is not true - some tests are going their own way. The class is also used as a container for variables that could be local - e.g. `inPointCloud` or `outPointCloud`. Additionally, `removeNonHitPoints` and `setUpPointCloud` could trivially become pure functions - which is preferred. Having those declarations would improve the readability of all tests. Consider the current:

```cpp
	int pointsCount = GetParam();

	setUpPointCloud(pointsCount, genCoord, genRandHit);

	// Print random seed for reproducibility
	fmt::print(stderr, "Compact Points Node Test random seed: {}\n", randomSeed);

	runGraphWithAssertions();

	removeNonHitPoints();

	TestPointCloud expectedPointCloud(pointFields, pointCoordValues.size());
	expectedPointCloud.setFieldValues<XYZ_F32>(pointCoordValues);
	expectedPointCloud.setFieldValues<IS_HIT_I32>(pointIsHitValues);

	outPointCloud = std::make_unique<TestPointCloud>(TestPointCloud::createFromNode(compactNodes.at(0), pointFields));

	EXPECT_EQ(*outPointCloud, expectedPointCloud);
```
vs alternative with pure functions and local declarations:
```cpp
	int pointsCount = GetParam();

	auto&& [inPointCloud, inNode] = setUpPointCloud(pointsCount, genCoord, genRandHit);

	// Print random seed for reproducibility
	fmt::print(stderr, "Compact Points Node Test random seed: {}\n", randomSeed);

	runGraphWithAssertions(inNode);

	auto&& expectedPointCloud = removeNonHitPoints(inPointCloud);
	auto&& outPointCloud = TestPointCloud::createFromNode(compactNodes.at(0), pointFields);

	EXPECT_EQ(*outPointCloud, expectedPointCloud);
```
IMHO the latter looks much clearer. This is, of course, only a rough example, leaving the details for you.

**Review by prybicki — APPROVED** [review]

**Pull request merged**